### PR TITLE
Feature/smhe 2266 thd single phase

### PR DESCRIPTION
**Pull request by harrymiddelburgetall**
Implemented the THD Fingerprint comandexecutor for single phase devices.
Values for phase 2 and 3 in THD Fingerprint response are now optional.

### CODE REVIEW

**Comment by harrymiddelburgetall** [comment]
✔https://jenkins.fdp.osgp.cloud/view/SMHE-cucumber/job/gxf-smhe-cucumber-build/266/

**Comment by harrymiddelburgetall** [comment]
✔ https://jenkins.fdp.osgp.cloud/view/SMHE-cucumber/job/gxf-smhe-cucumber-build/267/

File: osgp/platform/osgp-adapter-ws-smartmetering/src/test/java/org/opensmartgridplatform/adapter/ws/smartmetering/application/mapping/ThdFingerprintConverterTest.java [hunk 1]
```diff
@@ -75,8 +77,9 @@ void convertBundleWhenFingerprintNull() {
     assertThat(response.getThdFingerprint()).isNull();
   }
 
-  @Test
-  void convertSingle() {
+  @ParameterizedTest
+  @ValueSource(booleans = {true, false})
+  void convertSingle(final boolean polyphase) {
```
Comment (stefanermens) [inline osgp/platform/osgp-adapter-ws-smartmetering/src/test/java/org/opensmartgridplatform/adapter/ws/smartmetering/application/mapping/ThdFingerprintConverterTest.java:82]: `polyphase` is not used

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/misc/GetThdFingerprintCommandExecutor.java [hunk 2]
```diff
@@ -73,83 +98,73 @@ public GetThdFingerprintResponseDto execute(
...
-        currentL1,
-        currentL2,
-        currentL3,
-        fingerprintL1,
-        fingerprintL2,
-        fingerprintL3,
-        overLimitCounterL1,
-        overLimitCounterL2,
-        overLimitCounterL3);
+        resultMap.getOrDefault(THD_INSTANTANEOUS_CURRENT_L1, null),
```
Comment (stefanermens) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/misc/GetThdFingerprintCommandExecutor.java]: Is it possible to convert the result to the right type here? Then the constructor doesn't have to use parameters of type Object. I think it should be possible to do:
`(Integer) resultMap.getOrDefault(THD_INSTANTANEOUS_CURRENT_L1, null)`
If the getOrDefault returns null, the cast to Integer also returns null.

File: osgp/shared/osgp-dto/src/main/java/org/opensmartgridplatform/dto/valueobjects/smartmetering/GetThdFingerprintResponseDto.java [hunk 3]
```diff
@@ -11,54 +11,57 @@ public class GetThdFingerprintResponseDto extends ActionResponseDto {
...
-    this.thdCurrentOverLimitCounterL3 = thdCurrentOverLimitCounterL3;
+      final Object thdInstantaneousCurrentL1,
+      final Object thdInstantaneousCurrentL2,
+      final Object thdInstantaneousCurrentL3,
+      final Object thdInstantaneousCurrentFingerprintL1,
+      final Object thdInstantaneousCurrentFingerprintL2,
+      final Object thdInstantaneousCurrentFingerprintL3,
+      final Object thdCurrentOverLimitCounterL1,
+      final Object thdCurrentOverLimitCounterL2,
+      final Object thdCurrentOverLimitCounterL3) {
```
Comment (stefanermens) [inline osgp/shared/osgp-dto/src/main/java/org/opensmartgridplatform/dto/valueobjects/smartmetering/GetThdFingerprintResponseDto.java]: The types of the parameters can be Integer and List<Integer> if the cast to the right type is already done in the command executor.

File: osgp/shared/osgp-ws-smartmetering/src/main/resources/schemas/monitoring-ws-smartmetering.xsd [hunk 4]
```diff
@@ -688,15 +688,15 @@
...
-      <xsd:element name="ThdInstantaneousCurrentL1" type="xsd:int" />
-      <xsd:element name="ThdInstantaneousCurrentL2" type="xsd:int" />
-      <xsd:element name="ThdInstantaneousCurrentL3" type="xsd:int" />
-      <xsd:element name="ThdInstantaneousCurrentFingerprintL1" type="tns:FingerprintValues"/>
-      <xsd:element name="ThdInstantaneousCurrentFingerprintL2" type="tns:FingerprintValues"/>
-      <xsd:element name="ThdInstantaneousCurrentFingerprintL3" type="tns:FingerprintValues"/>
-      <xsd:element name="ThdCurrentOverLimitCounterL1" type="xsd:int" />
-      <xsd:element name="ThdCurrentOverLimitCounterL2" type="xsd:int" />
-      <xsd:element name="ThdCurrentOverLimitCounterL3" type="xsd:int" />
+      <xsd:element name="ThdInstantaneousCurrentL1" type="xsd:int" minOccurs="0" />
```
Comment (stefanermens) [inline osgp/shared/osgp-ws-smartmetering/src/main/resources/schemas/monitoring-ws-smartmetering.xsd]: For phase 1, it can be `minOccurs="1"`

**Review by stefanermens — CHANGES_REQUESTED** [review]

File: osgp/shared/osgp-dto/src/main/java/org/opensmartgridplatform/dto/valueobjects/smartmetering/GetThdFingerprintResponseDto.java (hunk 3 above)
Comment (harrymiddelburgetall) [reply osgp/shared/osgp-dto/src/main/java/org/opensmartgridplatform/dto/valueobjects/smartmetering/GetThdFingerprintResponseDto.java]: Refactored this. Conversion is done in command executor

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/misc/GetThdFingerprintCommandExecutor.java (hunk 2 above)
Comment (harrymiddelburgetall) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/misc/GetThdFingerprintCommandExecutor.java]: Refactored this. Conversion is done in command executor

File: osgp/platform/osgp-adapter-ws-smartmetering/src/test/java/org/opensmartgridplatform/adapter/ws/smartmetering/application/mapping/ThdFingerprintConverterTest.java (hunk 1 above)
Comment (harrymiddelburgetall) [reply osgp/platform/osgp-adapter-ws-smartmetering/src/test/java/org/opensmartgridplatform/adapter/ws/smartmetering/application/mapping/ThdFingerprintConverterTest.java:82]: Did not finish my intention to test. Now I did.

File: osgp/shared/osgp-ws-smartmetering/src/main/resources/schemas/monitoring-ws-smartmetering.xsd (hunk 4 above)
Comment (harrymiddelburgetall) [reply osgp/shared/osgp-ws-smartmetering/src/main/resources/schemas/monitoring-ws-smartmetering.xsd]: True, removed `minOccurs="0" ` 1 is default value

**Comment by harrymiddelburgetall** [comment]
⏳ https://jenkins.fdp.osgp.cloud/view/SMHE-cucumber/job/gxf-smhe-cucumber-build/269/

**Review by stefanermens — APPROVED** [review]

**Pull request merged**